### ISSUE TITLE
Adding a feature of "Hard" attention could be very useful

### ISSUE THREAD

**Issue by rakshith291** [tensorflow/tensorflow#50474]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.5
- Are you willing to contribute it (Yes/No): No

**Describe the feature and the current behavior/state.**
As of now we can get probability distribution after attention but would be quite useful if there is a way to use hard attention instead of a soft attention

**Will this change the current api? How?**
No
**Who will benefit with this feature?**
Could be quite useful while implementing ViT related architectures because using hard attention makes more sense for images instead of text 
**Any Other info.**

**Comment by Saduf2019** [comment]
@rakshith291 
Can you please share a use case to support your request.

**Comment by rakshith291** [comment]
@Saduf2019 
Use case : In case of dealing image patches as tokens in vision transformer it's quite helpful to have a hard attention where a given patch may have higher attention score with adjacent patch rather than all patches, hence hard attention could be quite useful

**Comment by rainwoodman** [comment]
@rakshit291 

Are you referring to the ViT example of Keras[1]?

The network uses a MultiHeadAttention layer which outputs probabilities generated by a softmax[2]. 

Because softmax preserves the ranking order, I think you may be able to insert a hardmax call after the output of MultiHeadAttention to achieve your desired effect. 

[1] https://keras.io/examples/vision/image_classification_with_vision_transformer/
[2] https://keras.io/api/layers/attention_layers/multi_head_attention/

**Comment by tilakrayal** [comment]
@rakshith291,
The usage of Softmax is needed in a neural network structure (as opposed to argmax) because it is continuous, and therefore differentiable. The argmax function is not 

https://math.stackexchange.com/questions/2354134/is-text-argmax-fx-y-a-differentiable-function-of-x

If you are interested in a sparse vector for output you could add another layer after softmax which would hard set the largest value from the softmax layer to 1 and all other values to zero. This would need to be done after training.

Also The "hardmax" function takes a vector and sets its largest element to 1, and all others to 0. The softmax function does almost the same thing, but it is continuous, and most machine learning techniques require this property to train neural networks, hence the "soft" modifier. Also please take a look at this link for the [reference](https://math.stackexchange.com/questions/1888141/why-is-the-softmax-function-called-that-way#:~:text=The%20%22hardmax%22%20function%20takes%20a,hence%20the%20%22soft%22%20modifier.). Thank you!

**Comment by tilakrayal** [comment]
Closing this as stale. Please reopen if this is still a valid request. Thank you!
